### DISCUSSION TITLE
simultaneous executions of terragrunt... is that a good idea?

### DISCUSSION THREAD

**Discussion by flixman** [gruntwork-io/terragrunt#3624] [Q&A]
We have a design for the deployment of our infrastructure, based on terragrunt. We have a pipeline that deploys infrastructure, and another pipeline that destroys it. We are wondering what will happen in case there are simultanous pipeline runs deploying infra, deploying and removing, and removing infra... will this be managed by terragrunt, like sequentialising the operations on the state files, or this situation is something we should better prevent?

**Comment by yhakbar** [comment]
Hey @flixman ,

You shouldn't run into any significant issues as long as you're taking advantage of state locking and provider caching for concurrent Terragrunt runs.

Terragrunt won't do anything to guarantee that one process applies before the other destroys, or vice versa, however.

Why exactly are you setting up one pipeline to destroy while the other deploys?

**Reply by flixman** [reply]
Thank you for your answer, @yhakbar! Our situation is: in terragrunt we have the state per component. We want the repository to model the landscape, and we'd like to just remove a folder and then terragrunt call destroy on the missing component. As we have not found the way to do it (although we are not experts with terragrunt), the option we have considered is to have a pipeline to perform destroy of the resources (that would also remove the state file and then check the removal of the resource in), while a CI pipeline would take care of the provisioning.

Is this the expected way to work? I'd expect this being a pretty common operation, so I am wondering if there is something wrong with our approach.

**Reply by yhakbar** [reply]
I gave an answer [here](https://github.com/gruntwork-io/terragrunt/discussions/3639#discussioncomment-11512042) that might help. Let me know if that answers your question.
